### PR TITLE
Add EnumValueResolver

### PR DESCRIPTION
**Pull request by ostrolucky**

### CODE REVIEW

**Comment by ostrolucky** [comment]
IMHO it's an edge case and 99% developers will rather rename variable in case of conflict than change wiring of this resolver. I can change the order if you wish though.

File: tests/Integration/Bridge/Symfony/ArgumentResolverTest.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface;
+
+class ArgumentResolverTest extends WebTestCase
+{
+    public function testValidEnumPassedInRequest()
+    {
+        if (!interface_exists(ArgumentValueResolverInterface::class)) {
+            $this->markTestSkipped(sprintf('"%s" is not present int this Symfony version', ArgumentValueResolverInterface::class));
```
Comment (ostrolucky) [inline tests/Integration/Bridge/Symfony/ArgumentResolverTest.php]: typo

**Comment by ogizanagi** [comment]
Oops. Sorry for the lack of response. 
Let's go as is. We cannot assume if the `GET` vs `POST` override is legit or not, so no exception.

**Comment by ostrolucky** [comment]
Good. I've fixed the typo

File: src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+class EnumValueResolver implements ArgumentValueResolverInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request, ArgumentMetadata $argument)
+    {
+        return is_subclass_of($argument->getType(), EnumInterface::class);
```
Comment (ogizanagi) [inline src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php]: `!$argument->isVariadic()` if we don't want to support variadics. Otherwise the resolve method should also handle arrays.

We must also check if the `$argument->getName()` is part of the attributes, query or request bags (using `has` on each).

File: src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve(Request $request, ArgumentMetadata $argument)
+    {
+        /** @var EnumInterface $enumClass */
+        $enumClass = $argument->getType();
+        $value = $request->get($argument->getName());
+
+        if (!$enumClass::accepts($value)) {
```
Comment (ogizanagi) [inline src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php]: `$value` could be `null` and expected if `$argument->isNullable()`, right?

File: src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php (hunk 2 above)
Comment (ostrolucky) [reply src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php]: What do you mean? Why do we need those has checks?

Comment (ogizanagi) [reply src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php]: See `RequestAttributeValueResolver` for instance. The support method should return `true` only if the argument can be found in the request. If it isn't, the default value provided on the controller will be used (if provided, otherwise an exception).

File: README.md [hunk 4]
```diff
@@ -537,6 +538,19 @@ abstract class AbstractEnumType extends Type
...
+<a href="https://symfony.com"><img src="https://img.shields.io/badge/Symfony-3.1%2B-green.svg?style=flat-square" title="Available for Symfony 3.1+" alt="Available for Symfony 3.1+" align="right"></a>
+
+There is an [argument value resolver](https://symfony.com/doc/current/controller/argument_value_resolver.html) provided,
+which seamlessly transforms HTTP request parameter into enum instance, just by typehinting the target enum in controller action.
+
+Simply register it inside the DIC configuration:
+```yaml
+services:
+    Elao\Enum\Bridge\Symfony\HttpKernel\Controller\ArgumentResolver\EnumValueResolver:
+        tags: [{ name: controller.argument_value_resolver }]
```
Comment (ogizanagi) [inline README.md]: Shouldn't the priority be set higher than 100 in order to be executed before `RequestAttributeValueResolver`, so you can still get a request attribute value as enum instead of raw value?

**Comment by ostrolucky** [comment]
failures I believe unrelated

File: composer.json [hunk 5]
```diff
@@ -28,14 +28,14 @@
     "prefer-stable": true,
     "minimum-stability": "dev",
     "require": {
-        "php": ">=7.0"
+        "php": ">=7.0",
+        "symfony/symfony": "^2.8"
```
Comment (chalasr) [inline composer.json]: should be reverted, this package can be used without any symfony component

File: src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve(Request $request, ArgumentMetadata $argument)
+    {
+        /** @var EnumInterface $enumClass */
+        $enumClass = $argument->getType();
+        $requestValue = $request->get($argument->getName());
```
Comment (ogizanagi) [inline src/Bridge/Symfony/HttpKernel/Controller/ArgumentResolver/EnumValueResolver.php]: Could just use:

```diff
- $requestValue = $request->get($argument->getName());
+ $requestValue = (array) $request->get($argument->getName());
```

and remove further if

**Comment by ostrolucky** [comment]
Let me do one more change

**Comment by ostrolucky** [comment]
ready

**Comment by ogizanagi** [comment]
@chalasr : Good to you too ? :)

**Comment by ogizanagi** [comment]
Thank you @ostrolucky.
